// iris-discord/src/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GatewayEvents = exports.GatewayIntents = exports.Constants = void 0;
class Constants {
}
exports.Constants = Constants;
Constants.opcodes = {
    DISPATCH: 0,
    HEARTBEAT: 1,
    IDENTIFY: 2,
    PRESENCE_UPDATE: 3,
    VOICE_STATE_UPDATE: 4,
    RESUME: 6,
    RECONNECT: 7,
    REQUEST_GUILD_MEMBERS: 8,
    INVALID_SESSION: 9,
    HELLO: 10,
    HEARTBEAT_ACK: 11
};
Constants.events = {
    ready: 'ready',
    messageCreate: 'messageCreate',
    messageUpdate: 'messageUpdate',
    messageDelete: 'messageDelete',
    messageDeleteBulk: 'messageDeleteBulk',
    messageReactionAdd: 'messageReactionAdd',
    messageReactionRemove: 'messageReactionRemove',
    messageReactionRemoveAll: 'messageReactionRemoveAll',
    messageReactionRemoveEmoji: 'messageReactionRemoveEmoji',
    guildCreate: 'guildCreate',
    guildUpdate: 'guildUpdate',
    guildDelete: 'guildDelete',
    guildBanAdd: 'guildBanAdd',
    guildBanRemove: 'guildBanRemove',
    guildEmojisUpdate: 'guildEmojisUpdate',
    guildIntegrationsUpdate: 'guildIntegrationsUpdate',
    guildMemberAdd: 'guildMemberAdd',
    guildMemberRemove: 'guildMemberRemove',
    guildMemberUpdate: 'guildMemberUpdate',
    guildMembersChunk: 'guildMembersChunk',
    guildRoleCreate: 'guildRoleCreate',
    guildRoleUpdate: 'guildRoleUpdate',
    guildRoleDelete: 'guildRoleDelete',
    inviteCreate: 'inviteCreate',
    inviteDelete: 'inviteDelete',
    presenceUpdate: 'presenceUpdate',
    typingStart: 'typingStart',
    userUpdate: 'userUpdate',
    voiceStateUpdate: 'voiceStateUpdate',
    voiceServerUpdate: 'voiceServerUpdate',
    webhooksUpdate: 'webhooksUpdate',
    interactionCreate: 'interactionCreate',
    shardReady: 'shardReady',
    shardResume: 'shardResume',
    shardDisconnect: 'shardDisconnect',
    shardError: 'shardError',
    shardReconnecting: 'shardReconnecting',
    invalidSession: 'invalidSession',
    raw: 'raw',
    debug: 'debug',
    warn: 'warn',
    error: 'error',
    rateLimit: 'rateLimit',
    shardRateLimit: 'shardRateLimit'
};
Constants.intents = {
    guilds: 1 << 0,
    guildMembers: 1 << 1,
    guildBans: 1 << 2,
    guildEmojis: 1 << 3,
    guildIntegrations: 1 << 4,
    guildWebhooks: 1 << 5,
    guildInvites: 1 << 6,
    guildVoiceStates: 1 << 7,
    guildPresences: 1 << 8,
    guildMessages: 1 << 9,
    guildMessageReactions: 1 << 10,
    guildMessageTyping: 1 << 11,
    directMessages: 1 << 12,
    directMessageReactions: 1 << 13,
    directMessageTyping: 1 << 14,
    messageContent: 1 << 15,
    guildScheduledEvents: 1 << 16,
    autoModerationConfiguration: 1 << 20,
    autoModerationExecution: 1 << 21,
};
Constants.messageTypes = {
    default: 0,
    recipientAdd: 1,
    recipientRemove: 2,
    call: 3,
    channelNameChange: 4,
    channelIconChange: 5,
    channelPinnedMessage: 6,
    guildMemberJoin: 7,
    userPremiumGuildSubscription: 8,
    userPremiumGuildSubscriptionTier1: 9,
    userPremiumGuildSubscriptionTier2: 10,
    userPremiumGuildSubscriptionTier3: 11,
    channelFollowAdd: 12,
    guildDiscoveryDisqualified: 14,
    guildDiscoveryRequalified: 15,
    reply: 19,
    applicationCommand: 20,
    threadStarterMessage: 21,
    guildInviteReminder: 22,
    contextMenuCommand: 23,
    threadStart: 24,
    threadUpdate: 25,
    threadDelete: 26,
    threadListSync: 27,
    threadMemberUpdate: 28,
    threadMembersUpdate: 29,
    guildScheduledEventCreate: 30,
    guildScheduledEventUpdate: 31,
    guildScheduledEventDelete: 32,
    channelThreadCreated: 33,
    replyThreadCreated: 34,
    guildScheduledEventCreateRequest: 35,
    guildScheduledEventUpdateRequest: 36,
    guildScheduledEventDeleteRequest: 37,
    threadMemberJoin: 38,
    threadMemberLeave: 39,
    threadMemberJoinRequest: 42,
    threadMemberLeaveRequest: 43,
    threadMemberUpdateRequest: 44,
    threadMembersUpdateRequest: 45,
    threadMemberJoinRequestApproved: 46,
    threadMemberJoinRequestDenied: 47,
    threadMemberLeaveRequestApproved: 48,
    threadMemberLeaveRequestDenied: 49,
    threadMemberUpdateRequestApproved: 50,
    threadMemberUpdateRequestDenied: 51,
    threadMembersUpdateRequestApproved: 52,
    threadMembersUpdateRequestDenied: 53,
};
Constants.interactionTypes = {
    chatInput: 1,
};
class GatewayIntents {
}
exports.GatewayIntents = GatewayIntents;
GatewayIntents.Guilds = 1 << 0;
GatewayIntents.GuildMembers = 1 << 1;
GatewayIntents.GuildBans = 1 << 2;
GatewayIntents.GuildEmojis = 1 << 3;
GatewayIntents.GuildIntegrations = 1 << 4;
GatewayIntents.GuildWebhooks = 1 << 5;
GatewayIntents.GuildInvites = 1 << 6;
GatewayIntents.GuildVoiceStates = 1 << 7;
GatewayIntents.GuildPresences = 1 << 8;
GatewayIntents.GuildMessages = 1 << 9;
GatewayIntents.GuildMessageReactions = 1 << 10;
GatewayIntents.GuildMessageTyping = 1 << 11;
GatewayIntents.DirectMessages = 1 << 12;
GatewayIntents.DirectMessageReactions = 1 << 13;
GatewayIntents.DirectMessageTyping = 1 << 14;
GatewayIntents.MessageContent = 1 << 15;
GatewayIntents.GuildScheduledEvents = 1 << 16;
GatewayIntents.AutoModerationConfiguration = 1 << 20;
GatewayIntents.AutoModerationExecution = 1 << 21;
class GatewayEvents {
}
exports.GatewayEvents = GatewayEvents;
GatewayEvents.Ready = 'ready';
GatewayEvents.MessageCreate = 'messageCreate';
GatewayEvents.MessageUpdate = 'messageUpdate';
GatewayEvents.MessageDelete = 'messageDelete';
GatewayEvents.MessageDeleteBulk = 'messageDeleteBulk';
GatewayEvents.MessageReactionAdd = 'messageReactionAdd';
GatewayEvents.MessageReactionRemove = 'messageReactionRemove';
GatewayEvents.MessageReactionRemoveAll = 'messageReactionRemoveAll';
GatewayEvents.MessageReactionRemoveEmoji = 'messageReactionRemoveEmoji';
GatewayEvents.GuildCreate = 'guildCreate';
GatewayEvents.GuildUpdate = 'guildUpdate';
GatewayEvents.GuildDelete = 'guildDelete';
GatewayEvents.GuildBanAdd = 'guildBanAdd';
GatewayEvents.GuildBanRemove = 'guildBanRemove';
GatewayEvents.GuildEmojisUpdate = 'guildEmojisUpdate';
GatewayEvents.GuildIntegrationsUpdate = 'guildIntegrationsUpdate';
GatewayEvents.GuildMemberAdd = 'guildMemberAdd';
GatewayEvents.GuildMemberRemove = 'guildMemberRemove';
GatewayEvents.GuildMemberUpdate = 'guildMemberUpdate';
GatewayEvents.GuildMembersChunk = 'guildMembersChunk';
GatewayEvents.GuildRoleCreate = 'guildRoleCreate';
GatewayEvents.GuildRoleUpdate = 'guildRoleUpdate';
GatewayEvents.GuildRoleDelete = 'guildRoleDelete';
GatewayEvents.InviteCreate = 'inviteCreate';
GatewayEvents.InviteDelete = 'inviteDelete';
GatewayEvents.PresenceUpdate = 'presenceUpdate';
GatewayEvents.TypingStart = 'typingStart';
GatewayEvents.UserUpdate = 'userUpdate';
GatewayEvents.VoiceStateUpdate = 'voiceStateUpdate';
GatewayEvents.VoiceServerUpdate = 'voiceServerUpdate';
GatewayEvents.WebhooksUpdate = 'webhooksUpdate';
GatewayEvents.InteractionCreate = 'interactionCreate';
GatewayEvents.ShardReady = 'shardReady';
GatewayEvents.ShardResume = 'shardResume';
GatewayEvents.ShardDisconnect = 'shardDisconnect';
GatewayEvents.ShardError = 'shardError';
GatewayEvents.ShardReconnecting = 'shardReconnecting';
GatewayEvents.InvalidSession = 'invalidSession';
GatewayEvents.Raw = 'raw';
GatewayEvents.Debug = 'debug';
GatewayEvents.Warn = 'warn';
GatewayEvents.Error = 'error';
GatewayEvents.RateLimit = 'rateLimit';
GatewayEvents.ShardRateLimit = 'shardRateLimit';
